lib/AST/ASTMangler.cpp

```diff
@@ -659,9 +659,34 @@ void ASTMangler::appendType(Type type) {
       appendFunctionType(cast<FunctionType>(tybase));
       return;
       
-    case TypeKind::SILBox:
-      appendType(cast<SILBoxType>(tybase)->getBoxedType());
-      return appendOperator("Xb");
+    case TypeKind::SILBox: {
+      auto box = cast<SILBoxType>(tybase);
+      auto layout = box->getLayout();
+      SmallVector<TupleTypeElt, 4> fieldsList;
+      for (auto &field : layout->getFields()) {
+        fieldsList.push_back(TupleTypeElt(field.getLoweredType()));
+      }
+      appendTypeList(TupleType::get(fieldsList, tybase->getASTContext())
+                       ->getCanonicalType());
+
+      if (auto sig = layout->getGenericSignature()) {
+        fieldsList.clear();
+        for (auto &arg : box->getGenericArgs()) {
+          fieldsList.push_back(TupleTypeElt(arg.getReplacement()));
+        }
+        appendTypeList(TupleType::get(fieldsList, tybase->getASTContext())
+                         ->getCanonicalType());
+        appendGenericSignature(sig);
+        appendOperator("XX");
+      } else {
+        appendOperator("Xx");
+      }
+
+      for (auto &field : layout->getFields())
+        appendOperator(field.isMutable() ? "m" : "i");
```

Comment:
Can you add the m/i list as an argument to the 'XX' operator, e.g. appendOperator('XX', miList).
It's a little bit more effort to store the mList in a separate char-array (like it's done in the SpecializationMangler), but it would give the correct mangling statistics output.